tests/groupRole.test.js

```diff
@@ -0,0 +1,1038 @@
+const assert = require('chai').assert;
+const cheerio = require('cheerio');
+const fetch = require('node-fetch').default;
+const path = require('path');
+const { step } = require('mocha-steps');
+
+const config = require('../lib/config');
+const sqldb = require('@prairielearn/postgres');
+const sql = sqldb.loadSqlEquiv(__filename);
+
+const helperServer = require('./helperServer');
+const { URLSearchParams } = require('url');
+
+let elemList;
+const locals = {};
+locals.helperClient = require('./helperClient');
+locals.siteUrl = 'http://localhost:' + config.serverPort;
+locals.baseUrl = locals.siteUrl + '/pl';
+locals.courseInstanceUrl = locals.baseUrl + '/course_instance/1';
+locals.assessmentsUrl = locals.courseInstanceUrl + '/assessments';
+locals.courseDir = path.join(__dirname, '..', 'testCourse');
+
+const storedConfig = {};
+
+describe('Test group based assessments with custom group roles from student side', function () {
+  this.timeout(20000);
+  before('set authenticated user', function () {
+    storedConfig.authUid = config.authUid;
+    storedConfig.authName = config.authName;
+    storedConfig.authUin = config.authUin;
+  });
+  before('set up testing server', helperServer.before(locals.courseDir));
+  after('shut down testing server', helperServer.after);
+  after('unset authenticated user', function () {
+    Object.assign(config, storedConfig);
+  });
+
+  describe('1. the database', function () {
+    step('should contain a group-based homework assessment with roles', async function () {
```

Comment:
Nesting `step` inside of `describe` is less than ideal: if a step inside a single `describe` block fails, only _that_ `describe` will be aborted, but the rest will keep executing. There are two things you could do here:

- Move all `step`s up to the root `describe` and get rid of the `describe('1. ...)` wrappers
- Instead of having multiple steps within each `describe`, merge the nested steps into a single step and use `step` instead of `describe to group. For example:

```ts
// Change from this:
describe('root', () => {
  describe('a', () => {
    step('1', () => console.log(1));
    step('2', () => console.log(2));
  });
});

// To this:
describe('root', () => {
  step('a', () => {
    console.log(1);
    console.log(2);
  })
});
```